### ISSUE TITLE
tf1.6 error: tf.contrib.ffmpeg.decode_video

### ISSUE THREAD

**Issue by XuChunling** [tensorflow/tensorflow#17533]
@yongtang  I use tensorflow 1.6on ubuntu and decodevideo.

`with tf.Session() as sess:
  movie_bin = tf.read_file('/home/xucl/app/data/bilibili/video/DongFangLieChe.mp4')
  movie = tf.contrib.ffmpeg.decode_video(movie_bin)
  movie_ev = movie.eval()
  print("****",len(movie_ev))`

but get an error

`2018-03-08 10:48:23.000491: F tensorflow/contrib/ffmpeg/default/ffmpeg_lib.cc:400] Non-OK-status: ReadInfoFile(stderr_filename, width, height, frames) status: Unknown: Not enough video info returned by FFmpeg [106, 0, 640, 3]Could not read FFmpeg stderr file: /tmp/tmp_file_tensorflow_3_cPNCGW.err
已放弃 (核心已转储)
`

**Comment by siekkine** [comment]
I got this error too (and a couple other ones). At least in my case, it was caused by a different version of ffmpeg installed in my system, which produces slightly different output. The underlying problem is that this functionality relies on parsing the ffmpeg output to extract metadata (nb of frames and frame dimensions) but the output may differ between versions as well as between video files. Perhaps a more reliable way would be to use ffprobe to get the metadata.

**Comment by yongtang** [comment]
I will take a look, thanks for the report.

**Comment by yongtang** [comment]
@XuChunling Can you provide the ffmpeg version you use if possible.

**Comment by XuChunling** [comment]
@yongtang the version of my ffmpeg

#################################################################

$ ffmpeg
ffmpeg version 2.4.3-1ubuntu1~trusty6 Copyright (c) 2000-2014 the FFmpeg developers
  built on Nov 22 2014 17:07:19 with gcc 4.8 (Ubuntu 4.8.2-19ubuntu1)
  configuration: --prefix=/usr --extra-version='1ubuntu1~trusty6' --build-suffix=-ffmpeg --toolchain=hardened --extra-cflags= --extra-cxxflags= --libdir=/usr/lib/x86_64-linux-gnu --shlibdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --enable-gpl --enable-shared --disable-stripping --enable-avresample --enable-avisynth --enable-fontconfig --enable-gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libflite --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libmodplug --enable-libmp3lame --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-librtmp --enable-libschroedinger --enable-libshine --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-opengl --enable-x11grab --enable-libxvid --enable-libx265 --enable-libdc1394 --enable-libiec61883 --enable-libzvbi --enable-libzmq --enable-frei0r --enable-libx264 --enable-libsoxr --enable-openal --enable-libopencv
  libavutil      54.  7.100 / 54.  7.100
  libavcodec     56.  1.100 / 56.  1.100
  libavformat    56.  4.101 / 56.  4.101
  libavdevice    56.  0.100 / 56.  0.100
  libavfilter     5.  1.100 /  5.  1.100
  libavresample   2.  1.  0 /  2.  1.  0
  libswscale      3.  0.100 /  3.  0.100
  libswresample   1.  1.100 /  1.  1.100
  libpostproc    53.  0.100 / 53.  0.100
Hyper fast Audio and Video encoder
usage: ffmpeg [options] [[infile options] -i infile]... {[outfile options] outfile}...

Use -h to get full help or, even better, run 'man ffmpeg'

**Comment by yongtang** [comment]
@XuChunling Is your system Ubuntu 14.04? Can you provide the details of the system, as was requested when you open the issue?

Ideally the availability of the video file would be great as well. But if that is not possible, can you run ffmpeg with your video and show the output (of stdout and stderr)?

**Comment by yongtang** [comment]
The following information will be really helpful:
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **TensorFlow installed from (source or binary)**:
- **TensorFlow version (use command below)**:
- **Python version**: 
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

**Comment by XuChunling** [comment]
Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
OS Platform and Distribution-->Linux Ubuntu 14.04
TensorFlow installed from-->binary
TensorFlow version -->1.6
Python version--> 3.5.5
GCC/Compiler version -->GCC 4.8.4
CUDA/cuDNN version-->cuda9/cuDNN7
GPU model and memory-->Tesla K20c/Tesla K20m


`import tensorflow as tf

with tf.Session() as sess:
  movie_bin = tf.read_file('/home/xucl/app/data/bilibili/video/DongFangLieChe.mp4')
  movie = tf.contrib.ffmpeg.decode_video(movie_bin)
  movie_ev = movie.eval()
  print("****",len(movie_ev))
`

**Comment by XuChunling** [comment]
@yongtang 

OS Platform and Distribution-->Linux Ubuntu 14.04
TensorFlow installed from-->binary
TensorFlow version -->1.6
Python version--> 3.5.5
GCC/Compiler version -->GCC 4.8.4
CUDA/cuDNN version-->cuda9/cuDNN7
GPU model and memory-->Tesla K20c/Tesla K20m
ffmpeg version-->2.4.3

code to reproduce:
`import tensorflow as tf

with tf.Session() as sess:
  movie_bin = tf.read_file('/home/xucl/app/data/bilibili/video/DongFangLieChe.mp4')
  movie = tf.contrib.ffmpeg.decode_video(movie_bin)
  movie_ev = movie.eval()
  print("****",len(movie_ev))`

get an error

`2018-03-08 10:48:23.000491: F tensorflow/contrib/ffmpeg/default/ffmpeg_lib.cc:400] Non-OK-status: ReadInfoFile(stderr_filename, width, height, frames) status: Unknown: Not enough video info returned by FFmpeg [106, 0, 640, 3]Could not read FFmpeg stderr file: /tmp/tmp_file_tensorflow_3_cPNCGW.err 已放弃 (核心已转储)`

but, when i use ffmpeg directly, It can be exactly run well

`$ ffmpeg -i /home/xucl/app/data/bilibili/video/DongFangLieChe.mp4 image-3%d.jpeg
ffmpeg version 2.4.3-1ubuntu1~trusty6 Copyright (c) 2000-2014 the FFmpeg developers
  built on Nov 22 2014 17:07:19 with gcc 4.8 (Ubuntu 4.8.2-19ubuntu1)
  configuration: --prefix=/usr --extra-version='1ubuntu1~trusty6' --build-suffix=-ffmpeg --toolchain=hardened --extra-cflags= --extra-cxxflags= --libdir=/usr/lib/x86_64-linux-gnu --shlibdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --enable-gpl --enable-shared --disable-stripping --enable-avresample --enable-avisynth --enable-fontconfig --enable-gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libflite --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libmodplug --enable-libmp3lame --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-librtmp --enable-libschroedinger --enable-libshine --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-opengl --enable-x11grab --enable-libxvid --enable-libx265 --enable-libdc1394 --enable-libiec61883 --enable-libzvbi --enable-libzmq --enable-frei0r --enable-libx264 --enable-libsoxr --enable-openal --enable-libopencv
  libavutil      54.  7.100 / 54.  7.100
  libavcodec     56.  1.100 / 56.  1.100
  libavformat    56.  4.101 / 56.  4.101
  libavdevice    56.  0.100 / 56.  0.100
  libavfilter     5.  1.100 /  5.  1.100
  libavresample   2.  1.  0 /  2.  1.  0
  libswscale      3.  0.100 /  3.  0.100
  libswresample   1.  1.100 /  1.  1.100
  libpostproc    53.  0.100 / 53.  0.100
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/home/xucl/app/data/bilibili/video/DongFangLieChe.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    encoder         : Lavf57.71.100
    description     : Packed by Bilibili XCoder v1.0(fixed_gap:False)
  Duration: 00:10:24.45, start: 0.000000, bitrate: 480 kb/s
    Stream #0:0(und): Video: h264 (High) (avc1 / 0x31637661), yuv420p, 640x360 [SAR 1:1 DAR 16:9], 408 kb/s, 20 fps, 20 tbr, 16k tbn, 40 tbc (default)
    Metadata:
      handler_name    : VideoHandler
    Stream #0:1(und): Audio: aac (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 65 kb/s (default)
    Metadata:
      handler_name    : SoundHandler
[swscaler @ 0x1857c40] deprecated pixel format used, make sure you did set range correctly
Output #0, image2, to 'image-3%d.jpeg':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    description     : Packed by Bilibili XCoder v1.0(fixed_gap:False)
    encoder         : Lavf56.4.101
    Stream #0:0(und): Video: mjpeg, yuvj420p, 640x360 [SAR 1:1 DAR 16:9], q=2-31, 200 kb/s, 20 fps, 20 tbn, 20 tbc (default)
    Metadata:
      handler_name    : VideoHandler
      encoder         : Lavc56.1.100 mjpeg
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> mjpeg (native))
Press [q] to stop, [?] for help
frame=12488 fps=227 q=24.8 Lsize=N/A time=00:10:24.40 bitrate=N/A
video:133937kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown
`

**Comment by XuChunling** [comment]
@siekkine Which version of the ffmpeg do you use?

**Comment by yongtang** [comment]
@XuChunling I was trying to install ffmpeg with Ubuntu 14.04 but it looks like I couldn't find it in the default repo (in Ubuntu 16.04 ffmpeg is available on the default repo). In order to reproduce exactly the issue you encountered, can you specify the method you install ffmpeg on your Ubuntu 14.04?

Also, in TensorFlow the command to execute ffmpeg for video docoding is:
```
ffmpeg -nostats -nostdin -i small.mp4 -f image2pipe -probesize 5000000 -loglevel info -hide_banner -vcodec rawvideo -pix_fmt rgb24 -y small.raw
```

where small.mp4 is the sample video and small.raw is the output file. Can you invoke the above command with the video files you have (DongFangLieChe.mp4) and post the output if possible?

**Comment by XuChunling** [comment]
@yongtang  the output

`$ ffmpeg -nostats -nostdin -i DongFangLieChe.mp4 -f image2pipe -probesize 5000000 -loglevel info -hide_banner -vcodec rawvideo -pix_fmt rgb24 -y small.raw
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'DongFangLieChe.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    encoder         : Lavf57.71.100
    description     : Packed by Bilibili XCoder v1.0(fixed_gap:False)
  Duration: 00:10:24.45, start: 0.000000, bitrate: 480 kb/s
    Stream #0:0(und): Video: h264 (High) (avc1 / 0x31637661), yuv420p, 640x360 [SAR 1:1 DAR 16:9], 408 kb/s, 20 fps, 20 tbr, 16k tbn, 40 tbc (default)
    Metadata:
      handler_name    : VideoHandler
    Stream #0:1(und): Audio: aac (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 65 kb/s (default)
    Metadata:
      handler_name    : SoundHandler
Output #0, image2pipe, to 'small.raw':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    description     : Packed by Bilibili XCoder v1.0(fixed_gap:False)
    encoder         : Lavf56.4.101
    Stream #0:0(und): Video: rawvideo (RGB[24] / 0x18424752), rgb24, 640x360 [SAR 1:1 DAR 16:9], q=2-31, 200 kb/s, 20 fps, 20 tbn, 20 tbc (default)
    Metadata:
      handler_name    : VideoHandler
      encoder         : Lavc56.1.100 rawvideo
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> rawvideo (native))
frame=12488 fps=216 q=0.0 Lsize= 8429400kB time=00:10:24.40 bitrate=110592.0kbits/s
video:8429400kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.000000%`

this is the output file

`$ ls
DongFangLieChe.mp4  ff_help  jpg  small.raw  XML`

**Comment by siekkine** [comment]
Unfortunately, I cannot check the ffmpeg version at the moment, but if I remember correctly, I pulled it recently from github and compiled it myself to enable CUDA support. Here are the parsing issues I met in my case:

- The aspect ratios ([SAR 1:1 DAR 16:9] above) outputted after the resolution confuses the code that parses height dimension. 

- Ordering of the output appears to differ between ffmpeg versions and, consequently, the parser code under "in_mapping" conditional may miss the "frame=  ". 

- With sufficiently long video, ffmpeg may output the line with "frame= " to stderr multiple times overwriting the previous output (using carriage return). The parser will find the first one whereas it should extract the last one (e.g., use rfind instead of find). 

I gave ffprobe a try and replaced the parsing functionality with it. It turns out that the frame count may sometimes be off (by one) unless one tells ffprobe explicitly to count them (-count_frames) but that slows down the whole thing because it requires an extra pass over the video file. So, that's not a perfect solution either. I also replaced the use of temp files with pipes to avoid extra I/O operations and implemented a way to sample every nth frame (pass select filter option to ffmpeg). My code is messy but I can share it with you if you want.

**Comment by yongtang** [comment]
@XuChunling @siekkine  I am able to find out where the issue is, will have a fix shortly.

By the way, when posting code blocks, you could use three back ticks ("```") at the begging line and end line to enclose the block so that it will render correctly:
```
$ ffmpeg -nostats -nostdin -i DongFangLieChe.mp4 -f image2pipe -probesize 5000000 -loglevel info -hide_banner -vcodec rawvideo -pix_fmt rgb24 -y small.raw
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'DongFangLieChe.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    encoder         : Lavf57.71.100
    description     : Packed by Bilibili XCoder v1.0(fixed_gap:False)
  Duration: 00:10:24.45, start: 0.000000, bitrate: 480 kb/s
    Stream #0:0(und): Video: h264 (High) (avc1 / 0x31637661), yuv420p, 640x360 [SAR 1:1 DAR 16:9], 408 kb/s, 20 fps, 20 tbr, 16k tbn, 40 tbc (default)
    Metadata:
      handler_name    : VideoHandler
    Stream #0:1(und): Audio: aac (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 65 kb/s (default)
    Metadata:
      handler_name    : SoundHandler
Output #0, image2pipe, to 'small.raw':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    description     : Packed by Bilibili XCoder v1.0(fixed_gap:False)
    encoder         : Lavf56.4.101
    Stream #0:0(und): Video: rawvideo (RGB[24] / 0x18424752), rgb24, 640x360 [SAR 1:1 DAR 16:9], q=2-31, 200 kb/s, 20 fps, 20 tbn, 20 tbc (default)
    Metadata:
      handler_name    : VideoHandler
      encoder         : Lavc56.1.100 rawvideo
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> rawvideo (native))
frame=12488 fps=216 q=0.0 Lsize= 8429400kB time=00:10:24.40 bitrate=110592.0kbits/s
video:8429400kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.000000%
```

**Comment by yongtang** [comment]
Added the fix in #17685.
